src/mesh/logical_location.hpp

```diff
@@ -0,0 +1,276 @@
+  if (lhs.morton() == rhs.morton()) return lhs.level() < rhs.level();
+  return lhs.morton() < rhs.morton();
+}
+
+inline bool operator>(const LogicalLocation &lhs, const LogicalLocation &rhs) {
+  if (lhs.morton() == rhs.morton()) return lhs.level() > rhs.level();
+  return lhs.morton() > rhs.morton();
+}
+
+inline bool operator==(const LogicalLocation &lhs, const LogicalLocation &rhs) {
+  return ((lhs.level() == rhs.level()) && (lhs.lx1() == rhs.lx1()) &&
+          (lhs.lx2() == rhs.lx2()) && (lhs.lx3() == rhs.lx3()));
+}
+
+struct block_ownership_t {
+ public:
+  const bool &operator()(int ox1, int ox2, int ox3) const {
+    return ownership[ox1 + 1][ox2 + 1][ox3 + 1];
+  }
+  bool &operator()(int ox1, int ox2, int ox3) {
+    return ownership[ox1 + 1][ox2 + 1][ox3 + 1];
+  }
+
+ private:
+  bool ownership[3][3][3];
+};
+
+inline block_ownership_t
+DetermineOwnership(const LogicalLocation &main_block,
+                   const std::set<LogicalLocation> &allowed_neighbors,
+                   RootGridInfo rg_info = RootGridInfo()) {
+  block_ownership_t main_owns;
+
+  auto ownership_less_than = [](const LogicalLocation &a, const LogicalLocation &b) {
+    // Ownership is first determined by block with the highest level, then by maximum
+    // Morton number this is reversed in precedence from the normal comparators where
+    // Morton number takes precedence
+    if (a.level() == b.level()) return a.morton() < b.morton();
+    return a.level() < b.level();
+  };
+
+  for (int ox1 : {-1, 0, 1}) {
+    for (int ox2 : {-1, 0, 1}) {
+      for (int ox3 : {-1, 0, 1}) {
+        auto possible_neighbors =
+            main_block.GetPossibleBlocksSurroundingTopologicalElement(ox1, ox2, ox3,
+                                                                      rg_info);
+
+        std::vector<LogicalLocation> actual_neighbors;
+        std::set_intersection(std::begin(allowed_neighbors), std::end(allowed_neighbors),
+                              std::begin(possible_neighbors),
+                              std::end(possible_neighbors),
+                              std::back_inserter(actual_neighbors));
+
+        auto max = std::max_element(std::begin(actual_neighbors),
+                                    std::end(actual_neighbors), ownership_less_than);
+        main_owns(ox1, ox2, ox3) =
+            (*max == main_block || ownership_less_than(*max, main_block) ||
+             actual_neighbors.size() == 0);
+      }
+    }
+  }
+  return main_owns;
+}
+
+} // namespace parthenon
+
+template <>
+struct std::hash<parthenon::LogicalLocation> {
+  std::size_t operator()(const parthenon::LogicalLocation &key) const noexcept {
+    // TODO(LFR): Think more carefully about what the best choice for this key is,
+    // probably the least significant sizeof(size_t) * 8 bits of the morton number
+    // with 3 * (level - 21) trailing bits removed.
+    return key.morton().bits[0];
```

Comment:
Similar to the question below: Why is it okay to just pick the first chunk of bits?

Bonus question: Do you foresee a case where the current choice of key causes issues? (I'm asking in terms of how important/relevant the `TODO` is)

Reply:
Well, this is just for hashing so we can choose whatever hash we would like I think (e.g. we could have this function always return one since there is no requirement for the hash to be unique, although this would of course give terrible performance). For less that 21 levels of refinement, the most significant bits should be a perfect hash, but for more refinement it is no longer perfect. Then, it seems like it might be better to use the bits described in the comment to minimize possible collisions. Since we are often just making `std::unordered_map`s of `LogicalLocations` in a small region of the grid, if we just used the most significant bits the hashes of all blocks around a corner that was at a refinement level > 22 would all collide if we just hashed with the most significant bits. Using the least significant bits would result in a locally perfect hash (most of the time).

Reply:
I am going to leave it as is for now.